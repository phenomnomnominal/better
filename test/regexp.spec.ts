import { createFixture } from './fixture';

describe('betterer', () => {
  it('should report the existence of RegExp matches', async () => {
    const { betterer } = await import('@betterer/betterer');

    const { logs, paths, readFile, cleanup, resolve, writeFile, testNames } = await createFixture('regexp', {
      '.betterer.js': `
const { regexp } = require('@betterer/regexp');

module.exports = {
  'regexp': () => regexp(/(\\/\\/\\s*HACK)/i).include('./src/**/*.ts')
};
    `
    });

    const configPaths = [paths.config];
    const resultsPath = paths.results;
    const indexPath = resolve('./src/index.ts');

    await writeFile(indexPath, `// HACK:`);

    const newTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(testNames(newTestRun.new)).toEqual(['regexp']);

    const sameTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(testNames(sameTestRun.same)).toEqual(['regexp']);

    await writeFile(indexPath, `// HACK:\n// HACK:`);

    const worseTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(testNames(worseTestRun.worse)).toEqual(['regexp']);

    const result = await readFile(resultsPath);

    expect(result).toMatchSnapshot();

    await writeFile(indexPath, ``);

    const betterTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(testNames(betterTestRun.better)).toEqual(['regexp']);

    const completedTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(testNames(completedTestRun.completed)).toEqual(['regexp']);

    expect(logs).toMatchSnapshot();

    await cleanup();
  });
});
