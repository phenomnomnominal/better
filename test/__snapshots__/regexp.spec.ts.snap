// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`betterer should report the existence of RegExp matches 1`] = `
"// BETTERER RESULTS V2.
//
// If this file contains merge conflicts, use \`betterer merge\` to automatically resolve them:
// https://phenomnomnominal.github.io/betterer/docs/results-file/#merge
//
exports[\`regexp\`] = {
  value: \`{
    \\"src/index.ts:4126639614\\": [
      [0, 0, 7, \\"RegExp match\\", \\"645651780\\"]
    ]
  }\`
};
"
`;

exports[`betterer should report the existence of RegExp matches 2`] = `
Array [
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 regexp: running \\"regexp\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ regexp: \\"regexp\\" got checked for the first time! (1 issue) 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" got checked for the first time! (1 issue) 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" got checked for the first time! (1 issue) 🎉

1 test got checked. 🤔
1 test got checked for the first time! 🎉
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 regexp: running \\"regexp\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ regexp: \\"regexp\\" stayed the same. (1 issue) 😐
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" stayed the same. (1 issue) 😐
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" stayed the same. (1 issue) 😐

1 test got checked. 🤔
1 test stayed the same. 😐
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 regexp: running \\"regexp\\"!
",
  "🌟 Betterer (0ms): 1 test running...
🔥 regexp: \\"regexp\\" got worse. (1 new issue, 2 total) 😔
",
  "🌟 Betterer (0ms): 1 test running...
🔥 regexp: \\"regexp\\" got worse. (1 new issue, 2 total) 😔

Error: \\"regexp\\" got worse. (1 new issue, 2 total) 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 regexp: \\"regexp\\" got worse. (1 new issue, 2 total) 😔
・ 1 existing issue in \\"<project>/fixtures/regexp/src/index.ts\\".
・ New issue in \\"<project>/fixtures/regexp/src/index.ts\\"!
・
・   <project>/fixtures/regexp/src/index.ts
・   1 | // HACK:
・ > 2 | // HACK:
・     | ^^^^^^^ RegExp match
・

Error: \\"regexp\\" got worse. (1 new issue, 2 total) 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 regexp: \\"regexp\\" got worse. (1 new issue, 2 total) 😔
・ 1 existing issue in \\"<project>/fixtures/regexp/src/index.ts\\".
・ New issue in \\"<project>/fixtures/regexp/src/index.ts\\"!
・
・   <project>/fixtures/regexp/src/index.ts
・   1 | // HACK:
・ > 2 | // HACK:
・     | ^^^^^^^ RegExp match
・

Error: \\"regexp\\" got worse. (1 new issue, 2 total) 😔

1 test got checked. 🤔
1 test got worse. 😔

You should try to fix the new issues! As a last resort, you can run \`betterer --update\` to force an update of the results file. 🆙
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 regexp: running \\"regexp\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ regexp: \\"regexp\\" met its goal! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" met its goal! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" met its goal! 🎉

1 test got checked. 🤔
1 test got better! 😍
1 test met its goal! 🎉
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 regexp: running \\"regexp\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ regexp: \\"regexp\\" has already met its goal! ✨
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" has already met its goal! ✨
",
  "🎉 Betterer (0ms): 1 test done!
✅ regexp: \\"regexp\\" has already met its goal! ✨

1 test got checked. 🤔
1 test met its goal! 🎉
",
]
`;
